### DISCUSSION TITLE
qt5-translations not handled by applocal.ps1

### DISCUSSION THREAD

**Discussion by Shatur** [microsoft/vcpkg#17867] [Ideas]
**Describe the bug**
[windeployqt](https://doc.qt.io/qt-5/windows-deployment.html) copies all Qt translations (from `share/qt5/translations/qt_<code>.qm`) to the folder with the specified executable file. This is necessary to make Qt-related strings translatable (for example, the standard cancel button text in dialogs). But `applocal.ps1` doesn't copy translations.

**To Reproduce**
Use `x_vcpkg_install_local_dependencies` on any Qt app target from the #14129.

**Comment by JackBoosY** [comment]
cc @Neumann-A

**Comment by Neumann-A** [comment]
this should be handled by `qtdeploy.ps1` in qt5-base but I don't actively support automatic deployment via vcpkg. I would always argue that one should use the functions of cmake + windeployqt for qt deployment.

**Comment by Shatur** [comment]
Automatic deployment very convenient for the apps that have a lot of dependencies. Copy everything manually using CMake would be insane.

Currently works fine for me, except the translations, I just copy them via CMake.

**Comment by Neumann-A** [comment]
> Copy everything manually using CMake would be insane.

I beg to differ
a) there is https://cmake.org/cmake/help/latest/command/file.html#get-runtime-dependencies
b) there is windeployqt which can be added as a custom command in cmake and do the work.
That is what I mean with "manually"

**Comment by Shatur** [comment]
>  a) there is https://cmake.org/cmake/help/latest/command/file.html#get-runtime-dependencies

I've tried this approach, but it copies many more libraries than required. Literally every library is copied, the application bundle size 4 times more than it needs :(

> b) there is windeployqt which can be added as a custom command in cmake and do the work.

Yes, it works as expected for Qt deps. But I switched to `applocal.ps1` to copy other dependencies because of the problem above with `GET_RUNTIME_DEPENDENCIES`.

So, adding translations to `applocal.ps1` won't hurt, this is a part of dependencies that used by `windeploqt`.

**Comment by Neumann-A** [comment]
>I've tried this approach, but it copies many more libraries than required. Literally every library is copied, the application bundle size 4 times more than it needs :(

That is simply a problem of not correctly setting it up. Use
```
  [PRE_INCLUDE_REGEXES [<regexes>...]]
  [PRE_EXCLUDE_REGEXES [<regexes>...]]
  [POST_INCLUDE_REGEXES [<regexes>...]]
  [POST_EXCLUDE_REGEXES [<regexes>...]]
```
to limit the amount of libraries it includes/tries to copies.

**Comment by Shatur** [comment]
>  [PRE_INCLUDE_REGEXES [<regexes>...]]
>  [PRE_EXCLUDE_REGEXES [<regexes>...]]
>  [POST_INCLUDE_REGEXES [<regexes>...]]
>  [POST_EXCLUDE_REGEXES [<regexes>...]]

I know about this, but most of the copied libraries had different names in my case and I would have a long list.
Not sure if manually setting this is a good idea.
What the problem with `applocal.ps1`?

**Comment by Neumann-A** [comment]
>I know about this, but most of the copied libraries had different names in my case and I would have a long list.

the trick is that POST regex can filter by paths/folders.

> What the problem with applocal.ps1?

Limits the script to always depend on vcpkg. Your build script probably knows a lot more about your build then the simple assumptions vcpkg does.

**Comment by Shatur** [comment]
> the trick is that POST regex can filter by paths/folders.

Paths where copied from? I didn't know about that. Will definitely try.

> Limits the script to always depend on vcpkg. Your build script probably knows a lot more about your build then the simple assumptions vcpkg does.

There is https://github.com/microsoft/vcpkg/pull/14243 that allows to pass `X_VCPKG_APPINSTALL_DEPS_INSTALL` at configure state and does all magic without any configuration from my side.
Of course, "native" CMake approach is better, but it just a bit hard to master :)

**Comment by Shatur** [comment]
>  b) there is windeployqt which can be added as a custom command in cmake and do the work.

@Neumann-A, you suggested to use `windeployqt`, but it doesn't work correctly if used from vcpkg folder:
1. It doesn't work for debug builds (because there are no debug versions of binaries next to it)
2. It does not copy translation files (because there are no translation files next to it)
3. It does not copy plugins.
4. It does not copy openssl libraries.
